Add FSD support for AOT

dsouzai opened the pull request:
Depends on ~(and therefore contains commits from)~ https://github.com/eclipse/openj9/pull/11553
Depends on ~(and therefore contains commits from)~ https://github.com/eclipse/openj9/pull/11879
Depends on https://github.com/eclipse/omr/pull/5814

This PR contains the changes necessary for enabling FSD for AOT:

- Adds Breakpoint Guard relocations
- Adds Feature Flag validation for FSD
- Improves FSD performance by no longer emitting method enter/exit hooks as these events are dealt with using involuntary OSR
- Modifies how a `TR_MethodPointer` relocation materializes a method pointer

Closes https://github.com/eclipse/openj9/issues/11376

dsouzai commented:
@mpirvu could you please review the control related changes
@mstoodle could you please review 
@vijaysun-omr fyi

mpirvu commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -8182,7 +8182,8 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
             bool aotCompilationReUpgradedToWarm = false;
-            if (that->_methodBeingCompiled->_useAotCompilation)
+            if (that->_methodBeingCompiled->_useAotCompilation
+                && !TR::Options::getCmdLineOptions()->getOption(TR_FullSpeedDebug))
```
I think a comment is warranted here explaining why we exclude FSD from doing warm AOT

dsouzai replied:
Added in https://github.com/eclipse/openj9/pull/11554/commits/b8505f40ed73b49c9551f75b3ef4385e2761a029

mpirvu reviewed (commented):
Overall it looks good to me. I have only some minor comments.

dsouzai commented on runtime/compiler/env/CHTable.cpp:
```diff
@@ -618,12 +618,17 @@ TR_CHTable::commitVirtualGuard(TR_VirtualGuard *info, List<TR_VirtualGuardSite>
          nopAssumptionIsValid = false;
-      ListIterator<TR_VirtualGuardSite> it(&sites);
-      for (TR_VirtualGuardSite *site = it.getFirst(); site; site = it.getNext())
+         }
+      else
          {
-         TR_PatchNOPedGuardSiteOnMethodBreakPoint
-            ::make(comp->fe(), comp->trPersistentMemory(), method, site->getLocation(), site->getDestination(), comp->getMetadataAssumptionList());
+         ListIterator<TR_VirtualGuardSite> it(&sites);
+         for (TR_VirtualGuardSite *site = it.getFirst(); site; site = it.getNext())
+            {
+            TR_PatchNOPedGuardSiteOnMethodBreakPoint
+               ::make(comp->fe(), comp->trPersistentMemory(), method, site->getLocation(), site->getDestination(), comp->getMetadataAssumptionList());
+            }
```
Added in https://github.com/eclipse/openj9/pull/11554/commits/23d89a50e80c81fa91d55122299f6c83a321bcc8

dsouzai commented on runtime/compiler/runtime/RelocationRuntime.cpp:
```diff
@@ -544,37 +544,16 @@ TR_RelocationRuntime::relocateAOTCodeAndData(U_8 *tempDataStart,
-         if (_comp->getOption(TR_FullSpeedDebug))
-            numExcptionRanges &= ~(J9_JIT_METADATA_WIDE_EXCEPTIONS | J9_JIT_METADATA_HAS_BYTECODE_PC);
-#endif /* defined(J9VM_OPT_JITSERVER) */
+         uint16_t numExcptionRanges = _exceptionTable->numExcptionRanges & 0x3fff;
```
Added in https://github.com/eclipse/openj9/pull/11554/commits/ed0d4c50525cf7dbe8ecc875b047909a9f27f011

mpirvu reviewed (approved):
LGTM

dsouzai commented:
Added ability to disable AOT FSD via env var; figured it's probably a good thing to have.

mstoodle commented on runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:
```diff
@@ -1150,6 +1150,18 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
          }
          break;
 
+      case TR_Breakpoint:
```
This commit deserves a non-empty message summarizing these changes :) .

mstoodle replied:
That's true of the first 3 commits, actually :) .

Sorry I left this comment hanging here! i got distracted after making this one comment and have been on vacation until today.

dsouzai replied:
Addressed in the latest force push.

mstoodle commented on runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:
```diff
@@ -546,12 +546,27 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
+          * in the binary template; at relo time, the inlined site index is
+          * sufficient to materialize the j9method pointer
+          */
+         if (j9method != inlinedJ9Method)
```
imo a better way to do this would be to create a new relocation record that does the inlining relocation as you want it, rather than using the existing `MethodPointer` relocation. That would avoid storing empty payload data as a flag to "process this one differently".

dsouzai replied:
I think that's in general a good idea. However, for the purposes of this PR, it might be best done in a separate PR later:

1. Adding a new relo type requires a change to OMR, so that's tedious; additionally there are uses of `TR_MethodPointer` in OMR so we would need to do a coordinate change where we're not only adding FSD but also modifying unrelated relo changes. Or we'd have to do that in a separate PR prior to these changes, in which case we might as well do it after.
2. `TR_MethodPointer` does nicely encapsulate the idea of "we have a pointer to a j9method". As such,  to determine in the code when we would use `TR_MethodPointer` and when we'd use the other might require us to check the inlinedSiteIndex every time we went to add the relo. This might be fine, but not something I want to do in this PR, especially because it should be done in a way that people shouldn't have to remember to always add the if.

Personally I don't mind the current approach ONLY because the "if this then that" logic is hidden away, and the two branches are still related to the same relocation. However, I do agree that it would be nice to have logical separation. So for the sake of not slowing down this PR, I'd prefer to just add it later on; in fact it could probably be bundled with the clean up that's needed in some of the inlined method relos where we use the cpIndex field for the SVM IDs heh.

mstoodle replied:
ok, I accept that the better approach would be a more invasive change. But I would prefer that better change. It might be possible to handle it transparently by having the MethodPointer external iterated relocation constructor automatically morph itself into a different relocation kind if it detects that the pointer matches the inlined site's j9method point. That doesn't change the fact that you would need an OMR change as well...

mstoodle replied:
if only the relocation framework were extensible....

mstoodle commented on runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:
```diff
@@ -546,12 +546,27 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
             j9method = reinterpret_cast<TR_OpaqueMethodBlock *>(aconstNode->getSymbolReference()->getSymbol()->castToStaticSymbol()->getStaticAddress());
 
-         TR_OpaqueClassBlock *j9class = fej9->getClassFromMethodBlock(j9method);
+         TR_InlinedCallSite & ics = comp->getInlinedCallSite(inlinedSiteIndex);
```
Please improve the commit summary "MethodPointer Relo Change" isn't the most generic you could have made it, but I think you could do better.

dsouzai replied:
Addressed in the latest force push.

mstoodle reviewed (changes requested):
Almost everything looks really good. I'm interested to hear your thoughts on my suggestion to add a new `MethodPointer` relocation type rather than storing empty payload data in the existing `MethodPointer` relocation.

dsouzai commented:
@mstoodle updated commit messages to be more descriptive.

mstoodle commented:
Thanks @dsouzai the commit messages are much more useful now.

dsouzai commented:
@mstoodle added new relo record, see https://github.com/eclipse/openj9/pull/11554/commits/75acda74099cd2581eecf411e14ecf3214d49e88 & https://github.com/eclipse/openj9/pull/11554/commits/822a1f745bf8203a1a4c3620570165f00f4d312f

dsouzai commented:
Linter failure is because of the dependent OMR change.

mstoodle commented on runtime/compiler/runtime/RelocationRecord.cpp:
```diff
@@ -5684,6 +5687,74 @@ TR_RelocationRecordMethodPointer::activatePointer(TR_RelocationRuntime *reloRunt
+
+         reloRuntime->comp()->setHasClassRedefinitionAssumptions();
+         }
+      }
```
would you please add an assert on the else here that
    `reloPrivatData->_clazz == reinterpret_cast<TR_OpaqueClassBlock *>(-1)`

dsouzai replied:
Made changes, see force push https://github.com/eclipse/openj9/compare/d11a837eafbc2a6a2406558cce5d4defaaa09ac9..4a69430d1d1603f0a4a8666a6ca4ed21fedce751

mstoodle reviewed (changes requested):
almost done, thanks for making the change I requested; I think it's much cleaner!

dsouzai commented:
Pushed again just to re-trigger the failed (due to infra issues) PR jobs.

@mstoodle the openj9-omr repo has the relo changes from OMR.

mstoodle commented:
Jenkins test sanity all jdk8,jdk11

dsouzai commented:
Worth noting, there's a bunch of additional jvmti tests that get run as part of `extended.functional` so might be worth running that too.

dsouzai commented:
PR Build failed because of 
```
Using context: Pull Request - OpenJ9
Pipeline_Build_Test_JDK8_x86-64_linux #2712 completed with status FAILURE (propagate: false to ignore)
Pipeline_Build_Test_JDK11_x86-64_linux_xl #2341 completed with status FAILURE (propagate: false to ignore)
Pipeline_Build_Test_JDK8_x86-64_windows #2792 completed with status FAILURE (propagate: false to ignore)
Finished: FAILURE
```
Not sure why those failing jobs aren't showing up here. At any rate, those three failing jobs are all infra related (see https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/5453/)

dsouzai commented:
Jenkins test extended.functional all jdk8,jdk11

dsouzai commented:
`Test_openjdk8_j9_extended.functional_x86-64_linux_Personal` failed because of https://github.com/eclipse/openj9/issues/8767.

dsouzai commented:
`Test_openjdk8_j9_extended.functional_x86-32_windows_Personal` is failing because of

```
[2021-02-22T21:47:37.994Z] Testing: rc018
[2021-02-22T21:47:37.994Z] Test start time: 2021/02/22 15:47:37 Central Standard Time
[2021-02-22T21:47:37.994Z] Running command: "F:/Users/jenkins/workspace/Grinder/openjdkbinary/j2sdk-image\\bin\\java"  -Xgcpolicy:optthruput -Xdebug -Xrunjdwp:transport=dt_socket,address=8888,server=y,onthrow=no.pkg.foo,launch=echo -Xjit:enableOSR,enableOSROnGuardFailure,count=1,disableAsyncCompilation -Xnocompressedrefs  -Xdump    -agentlib:jvmtitest=test:rc018 -cp "F:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\functional\cmdLineTests\jvmtitests\jvmtitest.jar;F:/Users/jenkins/workspace/Grinder/openjdk-tests/\TKG\lib\asm-all.jar" com.ibm.jvmti.tests.util.TestRunner
[2021-02-22T21:47:37.994Z] Time spent starting: 0 milliseconds
[2021-02-22T21:47:49.556Z] Time spent executing: 11796 milliseconds
[2021-02-22T21:47:49.556Z] Test result: FAILED
[2021-02-22T21:47:49.556Z] Output from test:
[2021-02-22T21:47:49.556Z]  [OUT] *** Testing [1/1]:	testReflectRedefineAtSameTime
[2021-02-22T21:47:49.556Z]  [OUT] starting reflect worker threads
[2021-02-22T21:47:49.556Z]  [OUT] starting to populate java heap
[2021-02-22T21:47:49.556Z] >> Success condition was not found: [Return code: 0]
```
I launched a grinder [1] and it's failing consistently, so I'll have to get on a win32 machine and debug this. No idea what the issue actually is (it's not a timeout...as far as I can tell anyway...) and why it's only failing on win32 🎉 .

dsouzai commented:
Not really sure what the issue is. Running the test manually results in the failure, quite consistently:
```
jenkins@win2012r2-x86-3 ~/dsouzai/fromConsole/openjdk-tests/TKG
$ "C:/cygwin64/home/jenkins/dsouzai/fromConsole/openjdkbinary/j2sdk-image\\bin\\java"  -Xgcpolicy:optthruput -Xdebug -Xrunjdwp:transport=dt_socket,address=8888,server=y,onthrow=no.pkg.foo,launch=echo '-Xjit:enableOSR,enableOSROnGuardFailure,count=1,disableAsyncCompilation,verbose={compilePerformance},vlog=vlog' -Xshareclasses:bootClassesOnly,nonFatal,silent,name=fsd_test_failure -Xnocompressedrefs  -Xdump   -agentlib:jvmtitest=test:rc018 -cp "C:/cygwin64/home/jenkins/dsouzai/fromConsole/openjdk-tests///..//jvmtest\functional\cmdLineTests\jvmtitests\jvmtitest.jar;C:/cygwin64/home/jenkins/dsouzai/fromConsole/openjdk-tests/\TKG\lib\asm-all.jar" com.ibm.jvmti.tests.util.TestRunner
*** Testing [1/1]:	testReflectRedefineAtSameTime
starting reflect worker threads
starting to populate java heap

jenkins@win2012r2-x86-3 ~/dsouzai/fromConsole/openjdk-tests/TKG
$ echo $?
127
```

But then running with `-Xaot:loadExclude={*}` also results in the same failure even though there's no AOT methods being compiled or loaded (verified with a verbose log): 
```
jenkins@win2012r2-x86-3 ~/dsouzai/fromConsole/openjdk-tests/TKG
$ grep 'AOT' vlog.20210223.152530.9876
#INFO:  AOT 
#INFO:  	SCC_stats_bytes: size=16776764 free=10440416 ROMClass=1498464 AOTCode=2576376 AOTData=224 JITHint=2052 JITProfile=50428
#INFO:  	SCC_stats_#:     ROMClasses=507 AOTMethods=1329 AOTDataEntries=1 AOTHints=99 AOTJitProfiles=511

```
However, it fails much much less frequently. To add to this, running with `-Xnoaot` doesn't result in a failure, even if I change the -Xjit counts to `count=1000,bcount=250` or `count=20,bcount=20`.

So, it's not bad AOT code that's causing the issue, but...enabling the feature seems to be a condition under which the issue will show up. Also, I don't even know what to make of this RC 127.

Therefore, this PR doesn't have any functional issues in and of itself. However, merging will definitely cause win32 nightly failures. @vijaysun-omr @mstoodle @pshipton thoughts/suggestions on how to move forward?

dsouzai commented:
Seems the common circumstance both in the run with AOT loads and with `-Xaot:loadExclude={*}` is that the issue happens shortly after redefinition interrupts the recompilation of `java/lang/Class.getDeclaredFields()[Ljava/lang/reflect/Field;`:
```
 (warm) Compiling java/lang/Class.getDeclaredFields()[Ljava/lang/reflect/Field;  OrdinaryMethod j9m=21F579A8 t=6018 compThreadID=0 memLimit=262144 KB freePhysicalMemory=6873 MB
#DISPATCH: Compiling java/lang/Class.getDeclaredFields()[Ljava/lang/reflect/Field; @ warm
#HD:  --load-- loader 2194A340, class 25E16000 : sun/reflect/GeneratedMethodAccessor8

#HD:  --init-- sun/reflect/GeneratedMethodAccessor8

#HD:  --updt-- sun/reflect/GeneratedMethodAccessor8

#HD:    extending sun/reflect/MethodAccessorImpl

#HD:    extending interface sun/reflect/MethodAccessor

#HD:  --load-- loader 21949F40, class 25EE0900 : com/ibm/jvmti/tests/redefineClasses/rc018_testRedefAndReflect_O1

#HK:  vmThread=219BCF00 hook jitClassesRedefined
#HD:  vmThread=219BCF00 hook jitClassesRedefined detail
  Invalidate all compilation requests#HK:  invalidateRequestsForUnloadedMethods class=00000000 vmThread=219BCF00 hotCodeReplacement=1
#HD:  vmThread=219BCF00 hook jitClassesRedefined detail
  Notify RAT on class old=25EE0600 fresh=25EE0900#HD:  vmThread=219BCF00 hook jitClassesRedefined detail
    Notify CHTable on method old=25EE051C fresh=25EE081C#HD:  vmThread=219BCF00 hook jitClassesRedefined detail
    Notify RAT on method old=25EE051C fresh=25EE081C#HD:  vmThread=219BCF00 hook jitClassesRedefined detail
  Notify CHTable on class old=25EE0600 fresh=25EE0900#HD:  vmThread=219BCF00 hook jitClassesRedefined finished
#DISPATCH: Failed to finalize compiled body for java/lang/Class.getDeclaredFields()[Ljava/lang/reflect/Field; @ warm
! (warm) java/lang/Class.getDeclaredFields()[Ljava/lang/reflect/Field; time=6139us compilationInterrupted memLimit=262144 KB freePhysicalMemory=6873 MB mem=[region=896 system=16384]KB
#CR:  219BCF00   Compile request OrdinaryMethod j9method=21F5BEC4 com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C optLevel=2
#CR:  219BCF00   Added entry 2197BD40 of weight 1 to comp queue. Now Q_SZ=1 weight=1
#DISPATCH: Compilation thread executing compile(): j9method=21F5BEC4 isAotLoad=0 canDoRelocatableCompile=0 eligibleForRelocatableCompile=0 isRemoteCompReq=0 _doNotUseAotCodeFromSharedCache=0 AOTfe=0 isDLT=0
 (cold) Compiling com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C  OrdinaryMethod j9m=21F5BEC4 t=6018 compThreadID=0 memLimit=262144 KB freePhysicalMemory=6873 MB
#DISPATCH: Compiling com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C @ cold
#DISPATCH: Successfully created compiled body [0x230F1E04-0x230F213B] for com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C @ cold
#DISPATCH: Creating metadata for com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C @ cold
#DISPATCH: Successfully created metadata [0x26D6F700] for com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C @ cold
+ (cold) com/ibm/jit/JITHelpers.getCharFromArrayByIndex(Ljava/lang/Object;I)C @ 0x230F1E04-0x230F213B OrdinaryMethod - Q_SZ=0 Q_SZI=0 QW=1 j9m=21F5BEC4 bcsz=158 sync OSR time=3221us mem=[region=1536 system=16384]KB compThreadID=0 CpuLoad=358%(44%avg) JvmCpu=358% compCPU=72%
#DISPATCH: CompilationEnd returning startPC=230F1E04  metadata=26D6F700
#CR:  219BCF00   Compile request OrdinaryMethod j9method=21F8D0E4 java/lang/StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder; optLevel=2
#CR:  219BCF00   Added entry 2197BD40 of weight 1 to comp queue. Now Q_SZ=1 weight=1
#DISPATCH: Compilation thread executing compile(): j9method=21F8D0E4 isAotLoad=0 canDoRelocatableCompile=0 eligibleForRelocatableCompile=0 isRemoteCompReq=0 _doNotUseAotCodeFromSharedCache=0 AOTfe=0 isDLT=0
 (cold) Compiling java/lang/StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder;  OrdinaryMethod j9m=21F8D0E4 t=6018 compThreadID=0 memLimit=262144 KB freePhysicalMemory=6873 MB
#DISPATCH: Compiling java/lang/StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder; @ cold
~                                                                                                                                                                                                                                             
~                                                                                                                                                                                                                                       
```
The fact that this EXACT circumstance was the cause of https://github.com/eclipse/openj9/pull/11879 makes me wonder if we do live in simulation... 

The weak excuse of a theory I can come up with that when the blocked app thread returns to the old body (that's now invalid and who's PC should've been patched) it does something that results in the RC 127. 🤷🏾‍♂️

dsouzai commented:
Jenkins test extended.functional all jdk8,jdk11

dsouzai commented:
Added a commit to disable recompilation if FSD & Sync Compilation. For justification:

```
An app thread that blocks on compilation releases VM Access. If class
redefinition occurs during a recompilation, the application thread
can no longer OSR out to the interpreter; it is forced to return to
the oldStartPC (to jump to a helper) which may not necessarily be
valid. Therefore, this commit disables recompilation if the JIT is
doing synchronous compilation in FSD mode.

In a normal run, compilations occur asynchronously. However, there
are circumstances where a recompilation must be done synchronously.
They have to do with method invalidation, specifically preexistence
and HCR. However, in FSD, preexistence is disabled, and HCR is
implemented using involuntary OSR. Therefore, the only
circumstance under which sync recompilations can occur under FSD is
when explicitly requested (via counts or disableAsyncCompilation).
```
Manually testing, this seems to have stopped the win32 RC127 issue. I don't know what the root cause is, so it's still possible to happen I suppose, but at the very least, this change makes it much less likely. It was mostly a guess that preventing recompilation would sort out the issue, which I guess it did (at least locally). Kicked off extended testing to verify here.

dsouzai commented:
> Test_openjdk8_j9_extended.functional_x86-32_windows_Personal — Build PASSED

🎉

dsouzai commented:
@mpirvu could you please review the change to disable recomp for FSD & Sync Compiles (https://github.com/eclipse/openj9/pull/11554/commits/3afec7da8a5a144511fa0193378d4cd00c2c0fb6)

dsouzai commented:
@mstoodle PR tests finally passing. Would appreciate a final review.

mstoodle reviewed (approved):
The AOT changes look good to me.

mpirvu commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -8254,7 +8254,7 @@ TR_J9SharedCacheVM::canMethodExitEventBeHooked()
 TR_J9SharedCacheVM::methodsCanBeInlinedEvenIfEventHooksEnabled()
    {
-   return true;
+   return !TR::Options::getCmdLineOptions()->getOption(TR_FullSpeedDebug);
```
We need to change the definition at JITServer which now reads:
```
bool
TR_J9SharedCacheServerVM::methodsCanBeInlinedEvenIfEventHooksEnabled()
   {
   return true;
   }
```
The complication is that we are returning the value from the global options, so I am proposing to add a `comp` parameter which can be used to retrieve the options: `comp->getOption(TR_FullSpeedDebug)`

dsouzai replied:
Addressed in https://github.com/eclipse/openj9/pull/11554/commits/6e15af8fa2a907dea4f78cda8bf953e2c0af41a5

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity all jdk11

mpirvu commented:
All tests passed except x86-64_linux_xl which as far as I can tell has an infra issue:
```
Still waiting to schedule task
‘basic-6p56l’ is offline
‘cent6-x64-3’ is reserved for jobs with matching label expression
‘cent6-x64-7’ is reserved for jobs with matching label expression
‘cent7-aarch64-1’ is reserved for jobs with matching label expression
‘cent7-ppcle-5’ is reserved for jobs with matching label expression
‘rh7-390-1’ is reserved for jobs with matching label expression
‘rh7-390-2’ is reserved for jobs with matching label expression
‘rh7-390-3’ is reserved for jobs with matching label expression
‘rh7-390-4’ is reserved for jobs with matching label expression
‘ub16-390-1’ is reserved for jobs with matching label expression
‘ub16-390-2’ is reserved for jobs with matching label expression
‘ub16p8j93’ is reserved for jobs with matching label expression
‘ub16x64j94’ is reserved for jobs with matching label expression
‘ub18-390-1’ is reserved for jobs with matching label expression
‘ub18-390-2’ is reserved for jobs with matching label expression
‘worker-1’ is reserved for jobs with matching label expression
‘worker-2’ is reserved for jobs with matching label expression
‘worker-3’ is reserved for jobs with matching label expression
```
TAP Extended Results does not show any error.
All in all this is ready to be merged.
